### PR TITLE
cmd: add service mode param

### LINKED ISSUE #5837: Make the service start with independent mode
## Development Task

- [ ] parse mode #5858
- [x] #5857
- [x] #5921
- [ ] #5933
- [ ] #5836
- [ ] #5956
- [ ] run independent mode with resource manager

### PR DESCRIPTION
**Pull request by lhy1024**
Signed-off-by: lhy1024 <admin@liudos.us>

### What problem does this PR solve?

Issue Number: Close #5837

### What is changed and how does it work?

1. cmd: split main.go #5915
2. decouple the dependency between server and mcs
3. run independent mode with resource manager

```commit-message
```

### Check List

Tests

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Code changes

- Has configuration change
- Has HTTP API interfaces changed (Don't forget to [add the declarative for the new API](https://github.com/tikv/pd/blob/master/docs/development.md#updating-api-documentation))
- Has persistent data change

Side effects

- Possible performance regression
- Increased code complexity
- Breaking backward compatibility

Related changes

- PR to update [`pingcap/docs`](https://github.com/pingcap/docs)/[`pingcap/docs-cn`](https://github.com/pingcap/docs-cn):
- PR to update [`pingcap/tiup`](https://github.com/pingcap/tiup):
- Need to cherry-pick to the release branch

### Release note

```release-note
Please add a release note.

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

If you don't think this PR needs a release note then fill it with None.
```

### CODE REVIEW

File: server/config/util.go [hunk 1]
```diff
@@ -111,3 +111,44 @@ func parseUrls(s string) ([]url.URL, error) {
...
+				return nil, errors.Errorf("duplicate service mode %s", item)
+			}
+		}
+		rets = append(rets, ret)
+	}
+	return rets, nil
+}
+
+func parseServiceMode(mode string) (ServiceMode, error) {
+	switch mode {
```
Comment (binshi-bing) [inline server/config/util.go:134]: switch strings.ToLower(mode) ?

File: server/config/util.go [hunk 2]
```diff
@@ -111,3 +111,44 @@ func parseUrls(s string) ([]url.URL, error) {
 
 	return urls, nil
 }
+
+func parseMode(modes string) ([]ServiceMode, error) {
+	items := strings.Split(modes, ",")
+	rets := make([]ServiceMode, 0)
+	for _, item := range items {
+		ret, err := parseServiceMode(item)
```
Comment (binshi-bing) [inline server/config/util.go:119]: Trim leading and trailing whitespace?

File: server/config/util.go [hunk 3]
```diff
@@ -111,3 +111,44 @@ func parseUrls(s string) ([]url.URL, error) {
 
 	return urls, nil
 }
+
+func parseMode(modes string) ([]ServiceMode, error) {
```
Comment (binshi-bing) [inline server/config/util.go:115]: Need unit test to show the valid and invalid input string of 'modes'.

**Review by binshi-bing — COMMENTED** [review]
left some comments

File: server/config/config.go [hunk 4]
```diff
@@ -201,6 +203,33 @@ func NewConfig() *Config {
...
+const (
+	APIService ServiceMode = iota
+	TSOService
+	ResourceManagerService
+	SchedulerService
+	ServiceModeCount
+)
+
+// String
+func (s ServiceMode) String() string {
```
Comment (binshi-bing) [inline server/config/config.go:218]: need to update the "services mode" section in "RFC: PD-Server supports multi-services mode" to make sure they are in sync.

File: server/config/config.go [hunk 5]
```diff
@@ -185,6 +186,7 @@ func NewConfig() *Config {
 	fs.StringVar(&cfg.AdvertiseClientUrls, "advertise-client-urls", "", "advertise url for client traffic (default '${client-urls}')")
 	fs.StringVar(&cfg.PeerUrls, "peer-urls", defaultPeerUrls, "url for peer traffic")
 	fs.StringVar(&cfg.AdvertisePeerUrls, "advertise-peer-urls", "", "advertise url for peer traffic (default '${peer-urls}')")
+	fs.StringVar(&cfg.ServiceModes, "service-modes", allModes(), "service modes support to start separately (default 'all')")
```
Comment (binshi-bing) [inline server/config/config.go:189]: It seems that we need another PR to support sub command

File: server/config/util.go (hunk 3 above)
Comment (lhy1024) [reply server/config/util.go]: yes, I will add them later.

File: cmd/pd-server/main.go [hunk 6]
```diff
@@ -123,21 +111,25 @@ func createServerWrapper(args []string) (context.Context, context.CancelFunc, ba
...
-		log.Fatal("join meet error", errs.ZapError(err))
-	}
-
-	// Creates server.
+	// todo: add more service mode and multi service mode
 	ctx, cancel := context.WithCancel(context.Background())
-	serviceBuilders := []server.HandlerBuilder{api.NewHandler, apiv2.NewV2Handler, swaggerserver.NewHandler, autoscaling.NewHandler}
-	serviceBuilders = append(serviceBuilders, dashboard.GetServiceBuilders()...)
-	svr, err := server.CreateServer(ctx, cfg, serviceBuilders...)
+	m, err := config.ParseServiceMode(cfg.ServiceModes)
```
Comment (binshi-bing) [inline cmd/pd-server/main.go:116]: In my understanding, based on our last discussion, config.ParseServiceMode will be switched to parse subcommand at the beginning of main(), correct?

Comment (lhy1024) [reply cmd/pd-server/main.go:116]: It will be did in another pr. This commit only sync with master

Comment (binshi-bing) [reply cmd/pd-server/main.go:116]: Thanks for the confirmation

**Review by binshi-bing — CHANGES_REQUESTED** [review]
@lhy1024 , could you please split the common part (e.g., move APIServiceGroup from server to pkg/utils/apiutil, registry refactor) required by other services, into a separate pr and check in first. The common part and tso mcs have higher priority than resource manager mcs. In this way, we can both work on tso mcs and resource manager mcs more efficiently.

cc @rleungx

**Review by binshi-bing — APPROVED** [review]

**Pull request closed without merging**